crates/aptos-runtimes/src/thread_manager.rs

```diff
@@ -0,0 +1,218 @@
+    Lazy::new(|| ThreadManager::new(ThreadManager::get_thread_config_strategy()));
+
+static THREAD_CONFIG_STRATEGY: OnceCell<ThreadConfigStrategy> = OnceCell::new();
+
+struct ThreadsConfig {
+    num_threads: usize,
+    on_thread_start: Box<dyn Fn() + Send + Sync + 'static>,
+}
+
+impl Default for ThreadsConfig {
+    fn default() -> Self {
+        Self::new_without_start_hook(num_cpus::get())
+    }
+}
+
+impl ThreadsConfig {
+    fn new_without_start_hook(num_threads: usize) -> Self {
+        Self::new(num_threads, || {})
+    }
+
+    fn new<F>(num_threads: usize, on_thread_start: F) -> Self
+    where
+        F: Fn() + Send + Sync + 'static,
+    {
+        Self {
+            num_threads,
+            on_thread_start: Box::new(on_thread_start),
+        }
+    }
+}
+
+#[derive(Default)]
+struct ThreadsConfigs {
+    tokio_config: ThreadsConfig,
+    rayon_config: ThreadsConfig,
+}
+
+struct Threads {
+    runtime: Runtime,
+    cpu_pool: ThreadPool,
+}
+
+pub struct ThreadManager {
+    exe_threads: Threads,
+    non_exe_threads: Threads,
+    io_threads: ThreadPool,
+}
+
+impl Default for ThreadManager {
+    fn default() -> Self {
+        Self::new(ThreadConfigStrategy::DefaultStrategy)
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum ThreadConfigStrategy {
+    DefaultStrategy,
+    #[cfg(target_os = "linux")]
+    PinExeThreadsToCores(usize),
+}
+
+// We probably want a "strategy trait" to define the methods below, and have different strategies
+// to implement them separately, but I'm lazy.
+impl ThreadManager {
+    pub fn get_thread_config_strategy() -> ThreadConfigStrategy {
+        match THREAD_CONFIG_STRATEGY.get() {
+            Some(strategy) => strategy.clone(),
+            None => ThreadConfigStrategy::DefaultStrategy,
+        }
+    }
+
+    pub fn set_thread_config_strategy(strategy: ThreadConfigStrategy) {
+        THREAD_CONFIG_STRATEGY
+            .set(strategy)
+            .expect("ThreadConfigStrategy can only be set once.");
+    }
+
+    pub fn get_exe_runtime(&self) -> Handle {
+        self.exe_threads.runtime.handle().clone()
+    }
+
+    pub fn get_exe_cpu_pool(&self) -> &ThreadPool {
+        &self.exe_threads.cpu_pool
+    }
+
+    pub fn get_non_exe_runtime(&self) -> Handle {
+        self.non_exe_threads.runtime.handle().clone()
+    }
+
+    pub fn get_non_exe_cpu_pool(&self) -> &ThreadPool {
+        &self.non_exe_threads.cpu_pool
+    }
+
+    pub fn get_io_pool(&self) -> &ThreadPool {
+        &self.io_threads
+    }
+
+    fn new(strategy: ThreadConfigStrategy) -> Self {
+        match strategy {
+            ThreadConfigStrategy::DefaultStrategy => {
+                let exe_threads = Self::create_threads(ThreadsConfigs::default(), "exe");
+                let non_exe_threads = Self::create_threads(ThreadsConfigs::default(), "non_exe");
+                let io_threads =
+                    Self::create_io_threads(ThreadsConfig::new_without_start_hook(64), "io");
```

Comment:
64 is very high, is this what we are using today?

Reply:
We are using multiple threadpools with 32 threads today. 

64 is not really high for real io if we have multiple disks, or using network disks.